### ISSUE TITLE
More Testing: Mocking is a Code Smell article is way too advanced

### ISSUE THREAD

**Issue by ssimx** [TheOdinProject/curriculum#28355]
### Checks

- [X] This is not a duplicate of an existing issue (please have a look through our [open issues list](https://github.com/TheOdinProject/curriculum/issues) to make sure)
- [X] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/.github/blob/main/CONTRIBUTING.md)
- [ ] Would you like to work on this issue?

### Describe your suggestion

I'm giving you my point of view as a beginner. This article is way too advanced, terminology used is nowhere to be found in previous lessons which wouldn't be a problem if it was only few terms and not whole (very long) article. Reading it feels like opening a book in the middle of it with no context provided and trying to understand what's going on. Also the jump from FunFunFunctions mock video to this article is night and day difference. I also struggled with some previous lessons but always knew where I was lacking so it was relatively easy to google it. This lesson I don't even know where to start, finding a course on this topic is my only option seems like. I wonder if others struggle as well or is it just me, feel free to discuss.

### Path

Node / JS

### Lesson Url

https://www.theodinproject.com/lessons/node-path-javascript-more-testing

### (Optional) Discord Name

vespucci

**Comment by incandesc3nce** [comment]
Agreed, the article felt like a really hard read and overly complicated for those not _yet_ familiar with React and Redux. But I think the article should stay, it gives some insight and tries to teach good coding techniques. Simply stating "glance through the article" and saying "it will have unknown terminology, but it is a good read" would be enough in my opinion. Maybe place it into additional resources .

**Comment by MaoShizhong** [comment]
I would not be opposed to moving the article in question to Additional Resources if someone can find a replacement that goes through the same core concepts, but presents them in a way that's more in line with the flow of the assignments. So there's less of a night/day jump in complexity, while retaining the core concepts discussed.

**Comment by MaoShizhong** [comment]
@TheOdinProject/javascript Any differing/additional thoughts on the above? Whether the assignment step itself could be moved to the AR section, or whether we only want to do so if we can get a replacement resource that goes through the same general concept in a more lesson-appropriate manner, or otherwise?

**Comment by JackBogart** [comment]
Bumping this up as I just made an issue on this, missed this when I went through the active issues! My original write up was in #28735 . 

To summarize my points were:

- This article is way too advanced
- The idea it's trying to convey isn't that complicated and could maybe be added to the lesson rather than being replaced by a different resource
- That idea being (rough summary): "too many mocks can be a sign of code smell as everything could tightly coupled if you need to mock often"

**Comment by YousifTheKind** [comment]
If there aren't articles to replace this one, perhaps it could be written in-house?

**Comment by cyphersept** [comment]
The lesson is very vague about what the "several really good points about program architecture and testing" from the article are. Any replacement material needs to be able to address those "several really good points" to a roughly equivalent quality.

Piggybacking off of [this discord discussion](https://discord.com/channels/505093832157691914/1179952911548174346/1290692811271704647) I'd summarize the points as:
1. If testing is making coding harder, not simpler, then you’re doing TDD wrong
2. Tests provide diminishing returns, with increasing clutter
3. What causes tight coupling, and how to avoid it
4. What are pure functions, and why they're ideal for testing
5. Code smell can tell you that there might be something wrong with your code
6. When is a mock bad

I submitted #29040 to discuss introducing 5 earlier in the curriculum in more accessible material, 4 is already addressed in the article linked in the pure functions section (which should probably get moved down to assignments), and 6 is already covered in the assignments via the FunFunFunction videos but tbh at this point a lesson on how to actually use mocking is probably more valuable than one telling you when not to use a feature you probably weren't going to use. So a replacement resource only needs to address 1-3.

- https://www.designite-tools.com/blog/understanding-testability-test-smells addresses only 1 and 3, but I really like this because it focuses on writing code that's easy to test, which has no overlaps with any other material other than the article to be replaced, and it's concise.
- [TDD, Where Did It All Go Wrong talk by Ian Cooper](https://youtu.be/EZ05e7EMOLM?si=ZyjDOsH3qMA4B4P0) comes highly recommended as a lesson on common TDD mistakes though I haven't watched it yet. It appears to cover all 3 and more, but it's an hour long video in a lesson that already has over an hour's worth of videos assigned. [This article](https://herbertograca.com/2018/08/27/distillation-of-tdd-where-did-it-all-go-wrong/) summarizes the contents.
- https://shiftmag.dev/test-driven-development-fail-and-success-1118/ covers 1 and 3. This also addresses writing testable code, though with less clarity than the first source, but the benefit of this article is that it's presented within the context of problem-solving as a newcomer to TDD which probably makes the cases it talks about easier to spot and apply
- Or we keep the original article, but specify what the "really good points" are and that the second half can be skipped.

**Comment by CouchofTomato** [comment]
@TheOdinProject/javascript Thoughts please.

**Comment by hoomanhorsey** [comment]
Just chiming in as I just finished this article and also found it really overwhelming.  Way over my head as a beginner.  It was so in-depth I just skimmed and scrolled through the last half as I couldn't really make sense of any of it. .  

On the quote:
"While it is quite in-depth, the following article contains several really good points about [program architecture and testing]" 

it's particularly frustrating because the article is so overwhelming I honestly have no idea what the really good points are. I can imagine a seasoned programmer nodding their head along with the article with the points it raises. But not as a beginner.

Meanwhile, the fun fun function material always leaves me with a smile on my face! So accessible and beautifully presented and explained.

Anyway, I feel like it's already acknowledged that the article may not be pitched at the right level of students.  But just wanted to add my feedback.

I do appreciate all the work that all of you put into this as well. Thanks

**Comment by wise-king-sullyman** [comment]
After reviewing the article, I think it can probably just be removed.

I agree that it is complex and honestly think it's out of scope for what a jr level developer really needs to know. I also don't see any key points in it that seem essential and aren't already covered otherwise in the curriculum.

**Comment by ZackHoang** [comment]
Mind if I take this on?

**Comment by JoshDevHub** [comment]
> Mind if I take this on?

@ZackHoang Still interested in taking it? I'll assign you if you want it.

**Comment by ZackHoang** [comment]
> > Mind if I take this on?
> 
> @ZackHoang Still interested in taking it? I'll assign you if you want it.

Yes please!

**Comment by JoshDevHub** [comment]
@ZackHoang Done. Let us know if you have any questions.
